### PR TITLE
[Memory] v2 Cap MLX buffer cache via measured warm-up overhead 

### PR DESCRIPTION
**Pull request by WindChimeRan**
built on #247 , to solve #234

memory_timeline_compare_f025

## Motivation
During sustained serving, MLX's buffer cache retains freed GPU buffers keyed by exact byte size. Because `total_tokens` varies each step, intermediate buffers are cached but never reused — they accumulate as dead weight until the OS reclaims memory. The previous mitigation (`mx.clear_cache()` every 50 requests) created a sawtooth pattern rather than bounding memory.

## Summary
```
load_model()                                       — weights only, no paged setup

determine_available_memory()                      ← paged setup now lives here
└─ scheduler_memory_reporting_mode()               ← config-only dispatch; backend may be None
   │                                                  (that's why the `backend is not None` guard
   │                                                   was removed — see comment 3)
   │
   └─ mode == "paged_attention_capacity":
      ├─ overhead = runner.profile_run()           ← NEW: measure + cap
      │    ├─ mx.clear_cache()
      │    ├─ cache_before = mx.get_cache_memory()          # ≈ 0 after clear
      │    ├─ dummy = mx.zeros((1, max_num_batched_tokens))
      │    ├─ mx.eval(_extract_logits(_forward_model(dummy))).  # use mlx_lm contiguous cache to ≈  page cache
      │    ├─ overhead = mx.get_cache_memory() - cache_before
      │    └─ mx.set_cache_limit(overhead)          ← THE KEY CALL
      │
      ├─ _setup_paged_attention(overhead=overhead) ← budget math + allocation, ONE site
      │    ├─ metal_limit = mx.device_info()[...]
      │    ├─ fraction    = VLLM_METAL_MEMORY_FRACTION 
      │    ├─ model_memory = _get_model_memory_usage()
      │    ├─ kv_budget   = metal_limit*fraction − model_memory − overhead
      │    │               (hybrid: −= max_num_seqs × linear_cache_bytes_per_slot)
      │    ├─ num_blocks  = kv_budget // per_block_bytes
      │    │   ├─ raise if kv_budget ≤ 0
      │    │   └─ raise if num_blocks < PAGED_ATTENTION_MIN_BLOCKS
      │    ├─ backend = _make_backend(runner, block_size)   # MHA / MLA / Hybrid
      │    ├─ backend.initialize(num_blocks)                # ← actual mx.zeros for KV tensors
      │    ├─ backend.patch_model(runner.model)
      │    └─ runner._paged_attention_backend = backend
      │
      └─ return backend.num_blocks() × block_size_bytes     # reports what was allocated

compile_or_warm_up_model()
└─ runner.warm_up()                                ← unchanged from main
   ├─ if STT:
   │    └─ stt_runtime_adapter.warm_up()
   └─ else (text / VLM):
      ├─ dummy (1,3) forward + mx.eval             # small dummy for Metal lib creation
      └─ if backend is not None:
         └─ backend.warm_up()                      # force kernel compile (still here)
```

Note that the dummy run is on mlx_lm's contiguous cache, just for the sake of profiling. The paged cache patching happens in the later stage. conceptually it's similar to the upstream vllm, which runs a stateless sdpa. 

Key difference (intentional design simplification) vllm is profilling the exact attn backend, while we are using contiguous cache as an approximation. 

## Minor change, coupled with this PR

*  hybrid paged attention is supported. remove unnecessary guard and comment
*  With the new design, we only guard stt model once, in `determine_available_memory`. Then `self.model_runner.should_setup_paged_attention` becomes dead code (removed)

## Benchmark (sonnet 1024+128, 100 prompts, concurrency 32)

| Metric | main | memory_cap_v2 | Δ |
|---|---:|---:|---:|
| Benchmark duration (s) | 196.34 | 196.65 | +0.16% |
| Output tok/s | 65.19 | 65.09 | −0.15% |
| Total tok/s | 580.78 | 579.85 | −0.16% |
| Mean TTFT (ms) | 11832.73 | 12147.85 | +2.66% |
| Mean TPOT (ms) | 381.45 | 379.75 | **−0.45%** |
| Median TPOT (ms) | 423.35 | 419.40 | **−0.93%** |
| P99 TPOT (ms) | 463.29 | 459.76 | **−0.76%** |
| Mean ITL (ms) | 381.45 | 379.75 | **−0.45%** |

**Verdict: no performance regression.** 

## Startup + configuration diff

| | main | memory_cap_v2 |
|---|---|---|
| Activation overhead source | 800 MB placeholder constant | `profile_run()` measurement |
| Reported overhead | `0.84 GB` | `0.75 GB (measured)` |
| KV budget | 11.37 GB | **11.46 GB** |
| num_blocks | 6195 | **6243** (+48) |
| Engine init time | 4.90 s | 10.46 s (+5.56 s) |

The +5.56 s startup cost is the profile forward pass at `max_num_batched_tokens=2048` tokens. One-time, paid at server boot. The measured value is ~50 MB lower than the placeholder on this model, so memory_cap_v2 recovers 48 additional cache blocks.

<details>
<summary>Full benchmark output</summary>

**main** (`790a40c`)
```
============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Maximum request concurrency:             32
Request rate configured (RPS):           10.00
Benchmark duration (s):                  196.34
Total input tokens:                      101230
Total generated tokens:                  12800
Request throughput (req/s):              0.51
Output token throughput (tok/s):         65.19
Peak output token throughput (tok/s):    256.00
Peak concurrent requests:                34.00
Total token throughput (tok/s):          580.78
---------------Time to First Token----------------
Mean TTFT (ms):                          11832.73
Median TTFT (ms):                        8163.57
P99 TTFT (ms):                           41863.18
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          381.45
Median TPOT (ms):                        423.35
P99 TPOT (ms):                           463.29
---------------Inter-token Latency----------------
Mean ITL (ms):                           381.45
Median ITL (ms):                         141.14
P99 ITL (ms):                            2819.84
==================================================
```

**memory_cap_v2** (`97fbce8`)
```
============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Maximum request concurrency:             32
Request rate configured (RPS):           10.00
Benchmark duration (s):                  196.65
Total input tokens:                      101230
Total generated tokens:                  12800
Request throughput (req/s):              0.51
Output token throughput (tok/s):         65.09
Peak output token throughput (tok/s):    256.00
Peak concurrent requests:                35.00
Total token throughput (tok/s):          579.85
---------------Time to First Token----------------
Mean TTFT (ms):                          12147.85
Median TTFT (ms):                        8168.39
P99 TTFT (ms):                           42664.31
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          379.75
Median TPOT (ms):                        419.40
P99 TPOT (ms):                           459.76
---------------Inter-token Latency----------------
Mean ITL (ms):                           379.75
Median ITL (ms):                         140.85
P99 ITL (ms):                            2770.22
==================================================
```
</details>

<details>
<summary>Benchmark config</summary>

- Model: `Qwen/Qwen3-0.6B`, `--max-model-len 2048`
- Dataset: sonnet (input 1024, output 128)
- Prompts: 100, request rate 10, max concurrency 32
- `VLLM_METAL_USE_PAGED_ATTENTION=1`, `VLLM_METAL_MEMORY_FRACTION=0.5`
- Hardware: Apple M1 Pro, 32 GB unified memory, macOS Darwin 25.4.0
- Commits: main `790a40c` · memory_cap_v2 `97fbce8`
</details>

### CODE REVIEW

File: vllm_metal/v1/model_runner.py [hunk 1]
```diff
@@ -799,6 +785,23 @@ def linear_cache_bytes_per_slot(self) -> int:
...
+        placeholder. ``mx.set_cache_limit`` prevents unbounded buffer-cache
+        growth during serving (issue #234).
+        """
+        warmup_len = self.scheduler_config.max_num_batched_tokens
+        mx.clear_cache()
+        cache_before = mx.get_cache_memory()
+        dummy_tokens = mx.zeros((1, warmup_len), dtype=mx.int32)
+        mx.eval(self._extract_logits(self._forward_model(dummy_tokens)))
+        overhead = max(0, mx.get_cache_memory() - cache_before)
+        mx.set_cache_limit(overhead)
```
Comment (LxYuan0420) [inline vllm_metal/v1/model_runner.py:551]: `profile_run()` is doing two things at once here: measuring overhead and setting cache policy. Can you clarify the contract a bit more? Right now it is also using `(1, max_num_batched_tokens)` as the probe, which looks like a heuristic rather than a clear runtime boundary.

Comment (LxYuan0420) [reply vllm_metal/v1/model_runner.py:551]: also what is the expected behavior when the measured overhead is 0 here?  mx.set_cache_limit(0) effectively disables the allocator cache?

File: vllm_metal/v1/model_runner.py [hunk 2]
```diff
@@ -348,7 +334,7 @@ def scheduler_memory_reporting_mode(
         """
         if self._is_stt:
             return "stt_nominal"
-        if paged_attention_enabled and self._paged_attention_backend is not None:
+        if paged_attention_enabled:
```
Comment (LxYuan0420) [inline vllm_metal/v1/model_runner.py:325]: Could you explain the ordering change here a bit more? The removed backend guard looks correct now that setup moved into `determine_available_memory()`, but it’s a bit subtle on first read.

**Review by LxYuan0420 — COMMENTED** [review]
Few things to clarify first

File: vllm_metal/v1/model_runner.py (hunk 1 above)
Comment (WindChimeRan) [reply vllm_metal/v1/model_runner.py:551]: Q1: At the high level we align with upstream vllm heuristics. Both bound the probe at `max_num_batched_tokens` . At the lower level we're strictly more conservative: upstream spreads those tokens across `max_num_seqs` requests, we run `(1, N)` which maximizes attention's O(L²) intermediate. 

Matching upstream's decomposition would require a more sophisticated attention-backend dispatch system than we have today.

Comment (WindChimeRan) [reply vllm_metal/v1/model_runner.py:551]: `max(0, ...)` This is overdefensive. I'll remove it. 

Total vllm memory: model weight + kv cache + overhead
```
    │     • usable_metal = metal_limit * fraction
    │     • kv_budget = usable_metal - model_memory - overhead   ← SUBTRACTION
```
Runtime error timing: 
* num_blocks < PAGED_ATTENTION_MIN_BLOCKS
*  kv_budget <= 0

File: vllm_metal/v1/model_runner.py (hunk 2 above)
Comment (WindChimeRan) [reply vllm_metal/v1/model_runner.py:325]: Clarification: see the updated diagram in the PR discription.

**Review by LxYuan0420 — APPROVED** [review]
